protocols/relay/src/priv_client/handler.rs

```diff
@@ -158,241 +174,303 @@ impl Handler {
+                // Fatal error occurred, thus handler is closing as quickly as possible.
+                // Transport is notified through dropping `to_listener`.
             }
+
+            let renewal = self.reservation.failed();
+
+            self.queued_events
+                .push_back(ConnectionHandlerEvent::NotifyBehaviour(
+                    Event::ReservationReqFailed {
+                        renewal,
+                        error: non_fatal_error,
+                    },
+                ));
+
+            return;
         }
-    }
+        // Try to process the error for a connection
+        let cmd = self.wait_for_connection_outbound_stream.pop_front().expect(
+            "got a stream error without a pending connection command or a reserve listener",
+        );
 
-    fn on_fully_negotiated_outbound(
-        &mut self,
-        FullyNegotiatedOutbound {
-            protocol: output,
-            info,
-        }: FullyNegotiatedOutbound<
-            <Self as ConnectionHandler>::OutboundProtocol,
-            <Self as ConnectionHandler>::OutboundOpenInfo,
-        >,
-    ) {
-        match (output, info) {
-            // Outbound reservation
-            (
-                outbound_hop::Output::Reservation {
-                    renewal_timeout,
-                    addrs,
-                    limit,
+        let non_fatal_error = match error {
+            StreamUpgradeError::Timeout => StreamUpgradeError::Timeout,
+            StreamUpgradeError::NegotiationFailed => StreamUpgradeError::NegotiationFailed,
+            StreamUpgradeError::Io(e) => {
+                self.pending_error = Some(StreamUpgradeError::Io(e));
+                return;
+            }
+            StreamUpgradeError::Apply(_) => unreachable!("should not update"),
+        };
+
+        let _ = cmd.send_back.send(Err(()));
+
+        self.queued_events
+            .push_back(ConnectionHandlerEvent::NotifyBehaviour(
+                Event::OutboundCircuitReqFailed {
+                    error: non_fatal_error,
                 },
-                OutboundOpenInfo::Reserve { to_listener },
-            ) => {
-                let event = self.reservation.accepted(
-                    renewal_timeout,
-                    addrs,
-                    to_listener,
-                    self.local_peer_id,
-                    limit,
-                );
+            ));
+    }
 
-                self.queued_events
-                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(event));
+    fn insert_to_deny_futs(&mut self, circuit: Circuit) {
+        let src_peer_id = circuit.src_peer_id();
+
+        if self.circuit_deny_futs.len() == MAX_NUMBER_DENYING_CIRCUIT
+            && !self.circuit_deny_futs.contains_key(&src_peer_id)
+        {
+            log::warn!(
+                "Dropping inbound circuit request to be denied from {:?} due to exceeding limit.",
+                src_peer_id
+            );
+        } else if self
```

Comment:
Please prefer an early return instead of `else if`, i.e. Use `return` in the above block and start with a new `if` after.